### DISCUSSION TITLE
Add 'flex-gap-*' utilities as a polyfill until general browser support is available

### DISCUSSION THREAD

**Discussion by kripod** [tailwindlabs/tailwindcss#2316] [Ideas]
The [`gap` property](https://tailwindcss.com/docs/gap) is very convenient to work with for grids. Unfortunately, browser support for the same property in flexbox contexts [isn't quite there yet](https://caniuse.com/#feat=flexbox-gap):

![Browser support data as viewable on CanIUse.com](https://user-images.githubusercontent.com/14854048/92283933-4cc1da80-ef01-11ea-84d4-4cf1d09351a3.png)

However, negative margins can be used for [simulating the same behavior](https://tailwindcss.com/components/flexbox-grids#column-spacing). It's a bit tedious to do right and every spacing gets doubled between elements:

```html
<div class="px-2">
  <div class="flex -mx-2">
    <div class="w-1/3 px-2">
      <div class="bg-gray-400 h-12"></div>
    </div>
    <div class="w-1/3 px-2">
      <div class="bg-gray-500 h-12"></div>
    </div>
    <div class="w-1/3 px-2">
      <div class="bg-gray-400 h-12"></div>
    </div>
  </div>
</div>
```

With a custom Tailwind plugin, `flex-gap`, `flex-gap-x` and `flex-gap-y` utilities could be added, replicating the behavior of gaps as specified in grids:

```js
/* tailwind.config.js */

const plugin = require("tailwindcss/plugin");

function half(value) {
  return value.replace(/\d+(.\d+)?/, (number) => number / 2);
}

module.exports = {
  plugins: [
    plugin(({ addUtilities, e, theme, variants }) => {
      Object.entries(theme("gap")).forEach(([key, value]) =>
        addUtilities(
          {
            [`.flex-gap-${e(key)}`]: {
              margin: `-${half(value)}`,
              "& > *": {
                margin: half(value),
              },
            },
            [`.flex-gap-x-${e(key)}`]: {
              marginRight: `-${half(value)}`,
              marginLeft: `-${half(value)}`,
              "& > *": {
                marginRight: half(value),
                marginLeft: half(value),
              },
            },
            [`.flex-gap-y-${e(key)}`]: {
              marginTop: `-${half(value)}`,
              marginBottom: `-${half(value)}`,
              "& > *": {
                marginTop: half(value),
                marginBottom: half(value),
              },
            },
          },
          variants("gap"),
        ),
      );
    }),
  ],
};
```

The HTML example above could be simplified into:

```html
<div class="px-2">
  <!-- Notice how the scale value is doubled below for consistency -->
  <div class="flex flex-gap-x-4">
    <div class="w-1/3">
      <div class="bg-gray-400 h-12"></div>
    </div>
    <div class="w-1/3">
      <div class="bg-gray-500 h-12"></div>
    </div>
    <div class="w-1/3">
      <div class="bg-gray-400 h-12"></div>
    </div>
  </div>
</div>
```

The best part is that all of the 'flex-gap' utilities may be replaced with just 'gap', once supported natively by the targeted browsers.

**Comment by taviks** [comment]
I learned that this approach can cause horizontal scrollbars to appear. 

In this Twitter thread (https://twitter.com/devongovett/status/1244679626162450432) learned from here (https://twitter.com/PetrOdut/status/1249264386943975426) that only using `margin-top` and `margin-left` fixes this, while still preserving the behavior of `gap`. Applying this to the plugin above gives us this:

```js
plugin(({ addUtilities, e, theme, variants }) => {
   Object.entries(theme('gap')).forEach(([key, value]) =>
      addUtilities(
         {
            [`.flex-gap-${e(key)}`]: {
               marginTop: `-${half(value)}`,
               marginLeft: `-${half(value)}`,
               '& > *': {
                  marginTop: half(value),
                  marginLeft: half(value)
               }
            }
         },
         variants('gap')
      )
   )
})
```

**Comment by adamwathan** [comment]
I think this is a great candidate for a third party plugin but I don't think it's a great fit for core, I try not to add stuff that will be made obsolete by browser features and would require a breaking change to remove. That's why the aspect ratio stuff is in a plugin instead of in core for example 👍🏻

**Comment by edlucas** [comment]
The margin plugin approach is problematic when you have nested `flex-gap` elements. In this case, the inner `flex` will be given both positive and negative margin styles, and only one set will be applied.
